Index: src/segment/Segment.cpp
===================================================================
--- src/segment/Segment.cpp	(revision 715)
+++ src/segment/Segment.cpp	(working copy)
@@ -2863,7 +2863,7 @@
 	//	To skip line-break slots; remember that the number of output slots equals the
 	//	number of gbbs plus initial and/or final line-break slots if any.
 	//	In this case the beginning of the first cluster indicates the first real glyph.
-	int isloutFirstReal = (m_dichwLim) ? ((m_fWsRtl)? visloutAfter[0] : visloutBefore[0]) : 0;
+	int isloutFirstReal = (m_dichwLim) ? visloutBefore[0] : 0;
 
 	//	Kludge to make assertions below easier:
 	if (m_dichwLim > 0)
@@ -2879,9 +2879,8 @@
 		// right-to-left situations.
 
 		if (prgiginfFirstOfCluster)
-			prgiginfFirstOfCluster[ich]
-				= LogicalToPhysicalSurface((m_fWsRtl) ? visloutAfter[ich] : visloutBefore[ich]);
-		int clusterStartPos = ((m_fWsRtl) ? visloutAfter[ich] : visloutBefore[ich]) - isloutFirstReal;
+			prgiginfFirstOfCluster[ich] = LogicalToPhysicalSurface(visloutBefore[ich]);
+		int clusterStartPos = (visloutBefore[ich] - isloutFirstReal);
 		if (pfClusterStart)
 			pfClusterStart[clusterStartPos] = true;
 
